juju/juju
# Add juju secrets backend and wire up to api facade

Add the backend secrets infrastructure for a juju secrets provider:
- secrets manager API facade
- juju secrets service provider
- mongo secrets store
- secrets URL parser

This PR provides a minimal implementation for create secret and get secret value - these now work end to end.
There's still a lot more todo but this is a start.

## QA steps

```
export JUJU_DEV_FEATURE_FLAGS=secrets
juju bootstrap lxd 
juju deploy ubuntu

juju exec --unit ubuntu/0 "secret-create password foo=bar hello=world"
secret://v1/f96c6f1b-9f81-41bf-890e-76a4f648b0fa/3547afab-0a6a-48be-87ca-1a5e3672a417/ubuntu.password

juju exec --unit ubuntu/0 "secret-get secret://v1/f96c6f1b-9f81-41bf-890e-76a4f648b0fa/3547afab-0a6a-48be-87ca-1a5e3672a417/ubuntu.password --format yaml"
foo: bar
hello: world
```

Also connect to mongo and inspect the `secretMetadata` and `secretValues` collections to see the secret info.
eg
```
juju:PRIMARY> db.secretValues.find().pretty()
{
        "_id" : "secret://v1/f96c6f1b-9f81-41bf-890e-76a4f648b0fa/3547afab-0a6a-48be-87ca-1a5e3672a417/ubuntu.password",
        "data" : {
                "foo" : "YmFy",
                "hello" : "d29ybGQ="
        },
        "txn-revno" : NumberLong(2),
        "txn-queue" : [
                "61161e533eeee028f25688e2_153e13c2"
        ]
}
```
